### PR TITLE
[EventHubs] Optimize body eager copying

### PR DESCRIPTION
**Pull request by danielmarbach**
Like https://github.com/Azure/azure-sdk-for-net/pull/31214

image

This benchmark used is

https://github.com/danielmarbach/MicroBenchmarks/commit/93d216e6d1489bfa5fde82fde8be6d317eedf3f4

# Contributing to the Azure SDK

Please see our [CONTRIBUTING.md](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md) if you are not familiar with contributing to this repository or have questions.

For specific information about pull request etiquette and best practices, see [this section](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#pull-request-etiquette-and-best-practices).

### CODE REVIEW

**Comment by danielmarbach** [comment]
@jsquire Josh preferred the common path to be "abstracted" so I did make the wish come true ;)

https://github.com/Azure/azure-sdk-for-net/pull/31214/commits/00ffa3e04cb40c041aafa26ce0cc58a63ed6ef24

let me know if you prefer this version as well

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/MessageBody.cs [hunk 1]
```diff
@@ -245,28 +259,49 @@ public EagerCopyingMessageBody(IEnumerable<Data> dataSegments)
...
 
-                var memory = _writer.GetMemory(dataToAppend.Length);
-                dataToAppend.CopyTo(memory);
+                // fields are lazy initialized to not occupy unnecessary memory when there are no data segments
+                _writer = length > 0 ? new ArrayBufferWriter<byte>(length) : new ArrayBufferWriter<byte>();
+                _segments = segments;
 
-                _writer.Advance(dataToAppend.Length);
-                _segments.Add(memory.Slice(0, dataToAppend.Length));
+                for (var i = 0; i < numberOfSegments; i++)
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Amqp/MessageBody.cs:297]: nit:  Can we call `i` something like `index` or `segmentIndex` and avoid the single-letter name?

**Review by jsquire — APPROVED** [review]
Thanks, @danielmarbach!   Just the one formatting ask and then a test run and we're good to go.

**Comment by jsquire** [comment]
@danielmarbach:  Looks like we're all good.  Ready to merge?

**Comment by danielmarbach** [comment]
All golden yes

**Pull request merged**